File: distro/customer-vps/host-bin/matrix-rotate-runtime-tokens.mjs
#!/opt/matrix/runtime/node/bin/node
import { createCipheriv, createDecipheriv, generateKeyPairSync, privateDecrypt, publicEncrypt, randomBytes, constants, createPublicKey } from 'node:crypto';
import { open, lstat, readFile, rename, unlink } from 'node:fs/promises';
import { dirname, resolve } from 'node:path';
import { pathToFileURL } from 'node:url';

const ENV_PATH = '/opt/matrix/env/host.env';
const KEY_PATH = '/opt/matrix/env/runtime-token-rotation.pem';
const TOKEN_KEYS = {
  sync: 'MATRIX_SYNC_RUNTIME_TOKEN',
  fundedAi: 'MATRIX_FUNDED_AI_RUNTIME_TOKEN',
  speech: 'MATRIX_PLATFORM_SPEECH_RUNTIME_TOKEN',
};
const MAX_FILE_BYTES = 65536;

function encoded(value) {
  if (typeof value !== 'string' || !/^[A-Za-z0-9+/]+={0,2}$/.test(value) || value.length > 8192) {
    throw new Error('Invalid rotation envelope');
  }
  return Buffer.from(value, 'base64');
}

function validPayload(payload) {
  if (!payload || typeof payload !== 'object' ||
      !/^[A-Za-z0-9_-]{1,128}$/.test(payload.machineId) ||
      !/^[a-z0-9-]{1,32}$/.test(payload.runtimeSlot) ||
      !Number.isSafeInteger(payload.epoch) || payload.epoch < 2 || payload.epoch > 2147483647 ||
      !payload.tokens || Object.keys(payload.tokens).sort().join(',') !== 'fundedAi,speech,sync' ||
      Object.values(payload.tokens).some((token) => typeof token !== 'string' || !/^[a-f0-9]{64}$/.test(token))) {
    throw new Error('Invalid rotation payload');
  }
  return payload;
}

export function encryptRuntimeTokenRotation(rawPayload, publicKey) {
  const payload = validPayload(rawPayload);
  const key = randomBytes(32);
  const nonce = randomBytes(12);
  const cipher = createCipheriv('aes-256-gcm', key, nonce);
  const ciphertext = Buffer.concat([cipher.update(JSON.stringify(payload)), cipher.final()]);
  const encryptedKey = publicEncrypt({ key: publicKey, padding: constants.RSA_PKCS1_OAEP_PADDING, oaepHash: 'sha256' }, key);
  return {
    version: 1,
    encryptedKey: encryptedKey.toString('base64'),
    nonce: nonce.toString('base64'),
    ciphertext: ciphertext.toString('base64'),
    tag: cipher.getAuthTag().toString('base64'),
  };
}

async function readRegular(path, maxBytes, rootOnly = false) {
  const info = await lstat(path);
  if (!info.isFile() || info.size > maxBytes || (info.mode & 0o022) || (rootOnly && (info.mode & 0o077))) {
    throw new Error('Unsafe rotation file');
  }
  if (rootOnly && typeof process.getuid === 'function' && process.getuid() === 0 && info.uid !== 0) {
    throw new Error('Unsafe rotation file owner');
  }
  return readFile(path);
}

function uniqueEnvValue(lines, key) {
  const matches = lines.filter((line) => line.startsWith(`${key}=`));
  if (matches.length !== 1) throw new Error('Invalid host environment');
  return matches[0].slice(key.length + 1);
}

export async function applyRuntimeTokenRotation({ envPath, keyPath, envelopePath }) {
  const [envBytes, privateKey, envelopeBytes] = await Promise.all([
    readRegular(envPath, MAX_FILE_BYTES),
    readRegular(keyPath, 8192, true),
    readRegular(envelopePath, MAX_FILE_BYTES),
  ]);
  const envelope = JSON.parse(envelopeBytes.toString('utf8'));
  if (envelope?.version !== 1 || Object.keys(envelope).sort().join(',') !== 'ciphertext,encryptedKey,nonce,tag,version') {
    throw new Error('Invalid rotation envelope');
  }
  const key = privateDecrypt({ key: privateKey, padding: constants.RSA_PKCS1_OAEP_PADDING, oaepHash: 'sha256' }, encoded(envelope.encryptedKey));
  const nonce = encoded(envelope.nonce);
  const tag = encoded(envelope.tag);
  if (key.length !== 32 || nonce.length !== 12 || tag.length !== 16) throw new Error('Invalid rotation envelope');
  const decipher = createDecipheriv('aes-256-gcm', key, nonce);
  decipher.setAuthTag(tag);
  const payload = validPayload(JSON.parse(Buffer.concat([decipher.update(encoded(envelope.ciphertext)), decipher.final()]).toString('utf8')));
  const env = envBytes.toString('utf8');
  const lines = env.split('\n');
  if (uniqueEnvValue(lines, 'MATRIX_MACHINE_ID') !== payload.machineId ||
      uniqueEnvValue(lines, 'MATRIX_RUNTIME_SLOT') !== payload.runtimeSlot) {
    throw new Error('Rotation target does not match this host');
  }
  const epochs = lines.filter((line) => line.startsWith('MATRIX_RUNTIME_TOKEN_EPOCH='));
  if (epochs.length > 1) throw new Error('Invalid host environment');
  const currentEpoch = epochs.length ? Number(epochs[0].slice('MATRIX_RUNTIME_TOKEN_EPOCH='.length)) : 1;
  if (!Number.isSafeInteger(currentEpoch) || payload.epoch !== currentEpoch + 1) {
    throw new Error('Unexpected runtime token epoch');
  }
  for (const keyName of Object.values(TOKEN_KEYS)) uniqueEnvValue(lines, keyName);
  const next = lines.map((line) => {
    for (const [field, keyName] of Object.entries(TOKEN_KEYS)) {
      if (line.startsWith(`${keyName}=`)) return `${keyName}=${payload.tokens[field]}`;
    }
    if (line.startsWith('MATRIX_RUNTIME_TOKEN_EPOCH=')) return `MATRIX_RUNTIME_TOKEN_EPOCH=${payload.epoch}`;
    return line;
  });
  if (!epochs.length) next.splice(next.at(-1) === '' ? next.length - 1 : next.length, 0, `MATRIX_RUNTIME_TOKEN_EPOCH=${payload.epoch}`);

  const destination = resolve(envPath);
  const directory = dirname(destination);
  const tempPath = `${destination}.rotation-${process.pid}-${randomBytes(8).toString('hex')}`;
  const original = await lstat(destination);
  let file;
  try {
    file = await open(tempPath, 'wx', original.mode & 0o777);
    await file.chown(original.uid, original.gid);
    await file.writeFile(next.join('\n'));
    await file.sync();
    await file.close();
    file = undefined;
    await rename(tempPath, destination);
    const directoryHandle = await open(directory, 'r');
    try { await directoryHandle.sync(); } finally { await directoryHandle.close(); }
  } finally {
    if (file) await file.close();
    await unlink(tempPath).catch((error) => { if (error.code !== 'ENOENT') throw error; });
  }
  return payload.epoch;
}

async function main() {
  if (typeof process.getuid !== 'function' || process.getuid() !== 0) throw new Error('Root is required');
  const action = process.argv[2];
  if (action === 'init' && process.argv.length === 3) {
    const { privateKey, publicKey } = generateKeyPairSync('rsa', { modulusLength: 3072 });
    const file = await open(KEY_PATH, 'wx', 0o600);
    try { await file.writeFile(privateKey.export({ type: 'pkcs8', format: 'pem' })); await file.sync(); }
    finally { await file.close(); }
    process.stdout.write(publicKey.export({ type: 'spki', format: 'pem' }));
    return;
  }
  if (action === 'public-key' && process.argv.length === 3) {
    const key = await readRegular(KEY_PATH, 8192, true);
    process.stdout.write(createPublicKey(key).export({ type: 'spki', format: 'pem' }));
    return;
  }
  if (action === 'apply' && process.argv.length === 4) {
    const epoch = await applyRuntimeTokenRotation({ envPath: ENV_PATH, keyPath: KEY_PATH, envelopePath: process.argv[3] });
    process.stdout.write(`Runtime token epoch ${epoch} installed. Restart dependent services.\n`);
    return;
  }
  throw new Error('Usage: matrix-rotate-runtime-tokens <init|public-key|apply ENCRYPTED_FILE>');
}

if (process.argv[1] && import.meta.url === pathToFileURL(resolve(process.argv[1])).href) {
  main().catch(() => { process.stderr.write('Runtime token rotation failed.\n'); process.exitCode = 1; });
}

File: distro/customer-vps/host-bin/matrix-rotate-runtime-tokens.py
#!/usr/bin/python3 -I
"""Root-only, encrypted, per-machine host token activation."""
import base64
import hashlib
import json
import os
import re
import stat
import sys
import tempfile

from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.asymmetric import padding, rsa
from cryptography.hazmat.primitives.ciphers.aead import AESGCM

ENV_PATH = '/opt/matrix/env/host.env'
KEY_PATH = '/opt/matrix/env/runtime-token-rotation.pem'
TOKEN_KEYS = {
    'sync': 'MATRIX_SYNC_RUNTIME_TOKEN',
    'fundedAi': 'MATRIX_FUNDED_AI_RUNTIME_TOKEN',
    'speech': 'MATRIX_PLATFORM_SPEECH_RUNTIME_TOKEN',
}


def read_regular(path, maximum, root_only=False):
    info = os.lstat(path)
    if not stat.S_ISREG(info.st_mode) or info.st_size > maximum or info.st_mode & 0o022:
        raise ValueError('Unsafe rotation file')
    if root_only and (info.st_mode & 0o077 or (os.geteuid() == 0 and info.st_uid != 0)):
        raise ValueError('Unsafe rotation key')
    with open(path, 'rb') as file:
        value = file.read(maximum + 1)
    if len(value) > maximum:
        raise ValueError('Unsafe rotation file')
    return value


def decode(value):
    if not isinstance(value, str) or len(value) > 8192 or not re.fullmatch(r'[A-Za-z0-9+/]+={0,2}', value):
        raise ValueError('Invalid rotation envelope')
    return base64.b64decode(value, validate=True)


def one_value(lines, key):
    values = [line[len(key) + 1:] for line in lines if line.startswith(key + '=')]
    if len(values) != 1:
        raise ValueError('Invalid host environment')
    return values[0]


def host_verifier_digest(env_path):
    lines = read_regular(env_path, 65536).decode('utf-8').split('\n')
    return hashlib.sha256(one_value(lines, 'UPGRADE_TOKEN').encode('utf-8')).hexdigest()


def apply_rotation(env_path, key_path, envelope_path):
    env_bytes = read_regular(env_path, 65536)
    private_key = serialization.load_pem_private_key(read_regular(key_path, 8192, True), password=None)
    envelope = json.loads(read_regular(envelope_path, 65536))
    if not isinstance(envelope, dict) or set(envelope) != {'version', 'encryptedKey', 'nonce', 'ciphertext', 'tag'} or envelope['version'] != 1:
        raise ValueError('Invalid rotation envelope')
    key = private_key.decrypt(decode(envelope['encryptedKey']), padding.OAEP(
        mgf=padding.MGF1(algorithm=hashes.SHA256()), algorithm=hashes.SHA256(), label=None,
    ))
    nonce = decode(envelope['nonce'])
    tag = decode(envelope['tag'])
    if len(key) != 32 or len(nonce) != 12 or len(tag) != 16:
        raise ValueError('Invalid rotation envelope')
    payload = json.loads(AESGCM(key).decrypt(nonce, decode(envelope['ciphertext']) + tag, None))
    tokens = payload.get('tokens') if isinstance(payload, dict) else None
    if (not isinstance(payload, dict) or not isinstance(payload.get('machineId'), str) or
            not isinstance(payload.get('runtimeSlot'), str) or
            not re.fullmatch(r'[A-Za-z0-9_-]{1,128}', payload['machineId']) or
            not re.fullmatch(r'[a-z0-9-]{1,32}', payload['runtimeSlot']) or
            type(payload.get('epoch')) is not int or not 2 <= payload['epoch'] <= 2147483647 or
            not isinstance(tokens, dict) or set(tokens) != set(TOKEN_KEYS) or
            any(not isinstance(token, str) or not re.fullmatch(r'[a-f0-9]{64}', token) for token in tokens.values())):
        raise ValueError('Invalid rotation payload')

    lines = env_bytes.decode('utf-8').split('\n')
    if (one_value(lines, 'MATRIX_MACHINE_ID') != payload['machineId'] or
            one_value(lines, 'MATRIX_RUNTIME_SLOT') != payload['runtimeSlot']):
        raise ValueError('Rotation target does not match this host')
    epochs = [line for line in lines if line.startswith('MATRIX_RUNTIME_TOKEN_EPOCH=')]
    if len(epochs) > 1:
        raise ValueError('Invalid host environment')
    current_epoch = int(epochs[0].split('=', 1)[1]) if epochs else 1
    if payload['epoch'] != current_epoch + 1:
        raise ValueError('Unexpected runtime token epoch')
    for key_name in TOKEN_KEYS.values():
        one_value(lines, key_name)
    replacements = {key_name: tokens[field] for field, key_name in TOKEN_KEYS.items()}
    replacements['MATRIX_RUNTIME_TOKEN_EPOCH'] = str(payload['epoch'])
    next_lines = [line.split('=', 1)[0] + '=' + replacements[line.split('=', 1)[0]]
                  if line.split('=', 1)[0] in replacements else line for line in lines]
    if not epochs:
        next_lines.insert(-1 if next_lines[-1] == '' else len(next_lines),
                          'MATRIX_RUNTIME_TOKEN_EPOCH=' + str(payload['epoch']))

    original = os.lstat(env_path)
    directory = os.path.dirname(os.path.abspath(env_path))
    descriptor, temporary = tempfile.mkstemp(prefix='.host.env.rotation-', dir=directory)
    try:
        with os.fdopen(descriptor, 'wb') as file:
            os.fchmod(file.fileno(), stat.S_IMODE(original.st_mode))
            os.fchown(file.fileno(), original.st_uid, original.st_gid)
            file.write('\n'.join(next_lines).encode('utf-8'))
            file.flush()
            os.fsync(file.fileno())
        os.replace(temporary, env_path)
        directory_fd = os.open(directory, os.O_RDONLY)
        try:
            os.fsync(directory_fd)
        finally:
            os.close(directory_fd)
    finally:
        if os.path.exists(temporary):
            os.unlink(temporary)
    return payload['epoch']


def main():
    if os.geteuid() != 0:
        raise ValueError('Root is required')
    if len(sys.argv) == 2 and sys.argv[1] == 'init':
        private_key = rsa.generate_private_key(public_exponent=65537, key_size=3072)
        private_pem = private_key.private_bytes(
            serialization.Encoding.PEM, serialization.PrivateFormat.PKCS8,
            serialization.NoEncryption(),
        )
        descriptor = os.open(KEY_PATH, os.O_WRONLY | os.O_CREAT | os.O_EXCL | os.O_NOFOLLOW, 0o600)
        with os.fdopen(descriptor, 'wb') as file:
            file.write(private_pem)
            file.flush()
            os.fsync(file.fileno())
        sys.stdout.buffer.write(private_key.public_key().public_bytes(
            serialization.Encoding.PEM, serialization.PublicFormat.SubjectPublicKeyInfo,
        ))
        return
    if len(sys.argv) == 2 and sys.argv[1] == 'public-key':
        private_key = serialization.load_pem_private_key(read_regular(KEY_PATH, 8192, True), password=None)
        sys.stdout.buffer.write(private_key.public_key().public_bytes(
            serialization.Encoding.PEM, serialization.PublicFormat.SubjectPublicKeyInfo,
        ))
        return
    if len(sys.argv) == 2 and sys.argv[1] == 'verifier-digest':
        print(host_verifier_digest(ENV_PATH))
        return
    if len(sys.argv) == 3 and sys.argv[1] == 'apply':
        epoch = apply_rotation(ENV_PATH, KEY_PATH, sys.argv[2])
        print(f'Runtime token epoch {epoch} installed. Restart dependent services.')
        return
    raise ValueError('Usage: matrix-rotate-runtime-tokens.py <init|public-key|verifier-digest|apply ENCRYPTED_FILE>')


if __name__ == '__main__':
    try:
        main()
    except Exception:
        print('Runtime token rotation failed.', file=sys.stderr)
        sys.exit(1)
